pkg/events/event_ringbuffer.go

```diff
@@ -0,0 +1,202 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package events
+
+import (
+	"sync"
+
+	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
+)
+
+type resultState int
+
+const (
+	resultOK resultState = iota
+	bufferEmpty
+	idNotFound
+)
+
+type eventRange struct {
+	start uint64
+	end   uint64
+}
+
+// eventRingBuffer A specialized circular buffer to store event objects.
+//
+// Unlike to regular circular buffers, existing entries are never directly removed and new entries can be added if the buffer is full.
+// In this case, the oldest entry is overwritten and can be collected by the GC.
+// Each event has an ID, however, this mapping is not stored directly. If needed, we calculate the id
+// of the event based on slice positions.
+//
+// Retrieving the records can be achieved with GetEventsFromID and GetRecentEntries.
+type eventRingBuffer struct {
+	events   []*si.EventRecord
+	capacity uint64
+	head     uint64
+	full     bool
+	id       uint64
+
+	sync.RWMutex
+}
+
+// Add adds an event to the ring buffer. If the buffer is full, the oldest element is overwritten.
+// This method never fails.
+func (e *eventRingBuffer) Add(event *si.EventRecord) {
+	e.Lock()
+	defer e.Unlock()
+
+	e.events[e.head] = event
+	if !e.full {
+		e.full = e.head == e.capacity-1
+	}
+	e.head = (e.head + 1) % e.capacity
+	e.id++
+}
+
+func (e *eventRingBuffer) GetEventsFromID(id uint64) ([]*si.EventRecord, resultState) {
+	e.RLock()
+	defer e.RUnlock()
+
+	pos, state := e.id2pos(id)
+	if state != resultOK {
+		return nil, state
+	}
+
+	if e.full && pos > e.head {
+		r1 := &eventRange{
+			start: pos,
+			end:   e.capacity,
+		}
+		r2 := &eventRange{
+			start: 0,
+			end:   e.head,
+		}
+		return e.getEntriesFromRanges(r1, r2), resultOK
+	}
+
+	end := e.head
+	if e.full && e.head == 0 {
+		// Special case: buffer is full and head is pointing at the beginning.
+		// Need to set explicitly to get [0..capacity] range because e.head
+		// never points to e.capacity.
+		end = e.capacity
+	}
+	return e.getEntriesFromRanges(&eventRange{
+		start: pos,
+		end:   end,
+	}, nil), resultOK
+}
+
+func (e *eventRingBuffer) GetRecentEntries(count uint64) ([]*si.EventRecord, uint64, resultState) {
```

Comment:
See above; I think we can remove this in favor of changing `GetEventsFromID()` slightly.

Reply:
I'll remove it if Wilfred acks the change.

Reply:
I think we can remove it. If we really need the functionality of `GetRecentEntries` then that would be a simple change in `GetEventsFromID` with `id == 0` as a special case here:
```
	if !idFound && id > 0 {
		return nil, lowest
	}
	if id == 0 {
		pos = e.head
	}
```